internal/controlplane/handlers_providers.go

```diff
@@ -156,6 +158,103 @@ func (_ *Server) ListProviderClasses(
+	entityCtx := engine.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+	providerName := entityCtx.Provider.Name
+
+	if providerName == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "provider name is required")
+	}
+
+	provider, err := s.providerStore.GetByNameInSpecificProject(ctx, projectID, providerName)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "provider not found")
+		}
+		return nil, status.Errorf(codes.Internal, "error getting provider: %v", err)
+	}
+
+	err = s.deleteProvider(ctx, provider, projectID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error deleting provider: %v", err)
+	}
+
+	return &minderv1.DeleteProviderResponse{
+		Name: providerName,
+	}, nil
+}
+
+// DeleteProviderByID deletes a provider by ID from a specific project.
+func (s *Server) DeleteProviderByID(
+	ctx context.Context,
+	in *minderv1.DeleteProviderByIDRequest,
+) (*minderv1.DeleteProviderByIDResponse, error) {
+	entityCtx := engine.EntityFromContext(ctx)
+	projectID := entityCtx.Project.ID
+
+	parsedProviderID, err := uuid.Parse(in.Id)
+	if err != nil {
+		return nil, util.UserVisibleError(codes.InvalidArgument, "invalid provider ID")
+	}
+
+	provider, err := s.providerStore.GetByID(ctx, parsedProviderID)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, util.UserVisibleError(codes.NotFound, "provider not found")
+		}
+		return nil, status.Errorf(codes.Internal, "error getting provider: %v", err)
+	}
+
+	err = s.deleteProvider(ctx, provider, projectID)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error deleting provider: %v", err)
+	}
+
+	return &minderv1.DeleteProviderByIDResponse{
+		Id: in.Id,
+	}, nil
+}
+
+func (s *Server) deleteProvider(ctx context.Context, provider *db.Provider, projectID uuid.UUID) error {
+	pbOpts := []providers.ProviderBuilderOption{
+		providers.WithProviderMetrics(s.provMt),
+		providers.WithRestClientCache(s.restClientCache),
+	}
+
+	p, err := providers.GetProviderBuilder(ctx, *provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
+	if err != nil {
+		return status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	}
+
+	// If the provider is a GitHub provider with a valid credential, delete all repositories associated with the provider
+	if p.Implements(db.ProviderTypeGithub) &&
+		providers.GetCredentialStateForProvider(ctx, *provider, s.store, s.cryptoEngine, &s.cfg.Provider) ==
+			provinfv1.CredentialStateSet {
+		client, err := p.GetGitHub()
+		if err != nil {
+			return status.Errorf(codes.Internal, "error creating github provider: %v", err)
+		}
+
+		// Delete all repositories associated with the provider and remove the webhooks
+		err = s.repos.DeleteRepositoriesByProvider(ctx, client, provider.Name, projectID)
+		if err != nil {
+			return status.Errorf(codes.Internal, "error deleting repositories: %v", err)
+		}
+	}
```

Comment:
This feels like it should be a part of the DeleteProvider implementation, rather than in `ControlPlane`.  e.g. GitHub providers should understand that they need to clean up webhooks, but other types of providers might have other types of cleanup they need to do.

(e.g. we talked about `DeleteProvider` for GHA un-installing the app from the org.)

Reply:
I had written it like that initially, but then `providerService` needed to use the `repositoryService` which seemed wrong. We could also call the `store` directly to delete the repositories, but then there will be duplicate logic in the services.

Reply:
I think it's totally reasonable for different services to call eachother - some domain concepts are at a higher level of abstraction than others, and thus delegate to the services for the lower-level domain concepts.

Reply:
It seems there were some recent changes that also now create an import cycle when I try that:
```
package github.com/stacklok/minder/internal/providers
	imports github.com/stacklok/minder/internal/repositories/github
	imports github.com/stacklok/minder/internal/logger
	imports github.com/stacklok/minder/internal/engine/actions/alert
	imports github.com/stacklok/minder/internal/engine/actions/alert/security_advisory
	imports github.com/stacklok/minder/internal/providers: import cycle not allowed
```

Reply:
I feel like we can resolve the cycle, but it's not something I want to spend time too much time on now. Can we live with this logic in the controlplane for now and do a followup PR to refactor?

Reply:
Yes, I'm okay with that.

Reply:
Looking at the import cycle, I'd probably break the `logger` --> `actions/alert` import path, but happy to do that as a follow-up.